### PR TITLE
Deem everything thrown as an error

### PR DESCRIPTION
**Pull request by sergiz**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

This is the first in line of PRs that will address issues outlined in https://github.com/fastify/fastify/issues/2860.

Firstly, I modified how kReplyIsError is tracked. The property is automatically set to false when reply.send() is used. This way there's no way that the reply would be considered an error longer than needed.

Secondly, I went through the code base and properly marked all replies as errors where appropriate. In the current version of Fastify there are huge inconsistencies of how non Errors are handled when thrown. Javascript supports throwing anything, so Fastify must take such cases into consideration. 
This PR ensures that you can safely throw anything in hooks or handlers regardless of them being sync or async. The error handler will be properly called in all such cases.

### CODE REVIEW

**Comment by mcollina** [comment]
Thanks for this PR! i think this is a good PR, however I am worrying a bit on using a positional argument for this, as it might get problematic to evolve this in the future (in case we need more options).

**Comment by jsumners** [comment]
I'm not really sold on the idea of changing the signature of `.send`. In my org we solve this by:

```js
module.exports = async function errorReply (fastify) {
  fastify.decorateReply('error', errorReply)
}

function errorReply (error, details) {
  // do stuff with the parameters
  this.code(/* whatever is appropriate */).send(theError)
}
```

I think this is solvable by a community/org plugin.

**Comment by airhorns** [comment]
IMO the "reply is error" idea should stay private to fastify. For someone new to the system, it's not really clear what `send("something", true)` means, you know? What does making a reply that is erroneous actually change about it? I think error conditions are something you don't generally opt into, but are something unexpected that are caught, and it seems like the original issue is more about catching new errors in the error handlers.

What about asking users to throw only `Error`s and just saying that throwing strings etc is unsupported? Yes you can throw anything in JS, but, it's kinda bad practice if you ask me, and Fastify gets a lot more information about an error if it is a real error object. You can serialize it differently, you can log the stack to the console, etc etc, shouldn't we be encouraging that?

**Comment by sergiz** [comment]
Thanks for all your inputs. I modified the PR now. Signature of reply.send() is no longer modified. I modified the way that kReplyIsError is tracked instead. It will automatically turn itself to false when reply.send() is used.

**Comment by jsumners** [comment]
> I modified the way that kReplyIsError is tracked instead. It will automatically turn itself to false when reply.send() is used.

Why? What does this solve?

**Review by mcollina — APPROVED** [review]
lgtm, I'm ok with the current implementation

**Comment by sergiz** [comment]
> 
> 
> > I modified the way that kReplyIsError is tracked instead. It will automatically turn itself to false when reply.send() is used.
> 
> Why? What does this solve?

Makes it easy to know what is going and saves time since you don't have to think of all cases where you have to manually set kReplyIsError to false.

File: lib/contentTypeParser.js [hunk 1]
```diff
@@ -13,7 +13,8 @@ const {
   kBodyLimit,
   kRequestPayloadStream,
   kState,
-  kTestInternals
+  kTestInternals,
+  kReplyisError
```
Comment (delvedor) [inline lib/contentTypeParser.js:17]: Typo, the `i` should be uppercase. I guess the test are not covering this.

**Review by delvedor — CHANGES_REQUESTED** [review]
I agree with @airhorns, throwing anything but errors it's wrong and should not be supported. Maybe this is a breaking change we should do in v4.

Can you check the status code in the test as well?

**Comment by mcollina** [comment]
> I agree with @airhorns, throwing anything but errors it's wrong and should not be supported. Maybe this is a breaking change we should do in v4.

This is already targeting `next`/v4.

**Comment by mcollina** [comment]
I think the reason we should ship this is `Jest`. You probably might not know that `Jest` overrides all globals (including `Error`) for each test run. However Node.js core globals cannot be tampered with, so errors thrown by Node.js core fails a `instanceof Error` check.

In other terms, we cannot safely assume that what is thrown is an instance of `Error`.

**Comment by climba03003** [comment]
> 
> 
> I think the reason we should ship this is `Jest`. You probably might not know that `Jest` overrides all globals (including `Error`) for each test run. However Node.js core globals cannot be tampered with, so errors thrown by Node.js core fails a `instanceof Error` check.
> 
> In other terms, we cannot safely assume that what is thrown is an instance of `Error`.

I do not know if it is true or not, `Jest` is actually running all test inside `vm.Context`. The script running inside should use the same set of global environment. If the provided `Error` is the same, it should not be a problem when using `Jest` as test framework. At least, I cannot think of a test-case that fail the `err instanceOf Error` assertion.

**Comment by mcollina** [comment]
> I missed the important point, error thrown by Node.js core.

Exactly that. Note that all the usage of `require('assert')` suffer from this problem.

**Comment by delvedor** [comment]
@mcollina could this be solved by catching whatever Jest is throwing instead of listen for everything?

**Comment by mcollina** [comment]
> @mcollina could this be solved by catching whatever Jest is throwing instead of listen for everything?

unfortunately not. We do not have access to the primordials from Node.js core.

**Comment by jsumners** [comment]
So is this a revert of https://github.com/fastify/fastify/pull/2134 before we ever even shipped it?

**Comment by mcollina** [comment]
> So is this a revert of https://github.com/fastify/fastify/pull/2134 before we ever even shipped it?

Why do you say so? I think it moves forward into that same direction.

**Comment by jsumners** [comment]
> > So is this a revert of #2134 before we ever even shipped it?
> 
> Why do you say so? I think it moves forward into that same direction.

I see. 🤔 I suppose if someone is going to `throw 'a string'` it's really on them to have an error handler that can detect that.

**Comment by mcollina** [comment]
> I see. 🤔 I suppose if someone is going to throw 'a string' it's really on them to have an error handler that can detect that.

Essentially. This should probably be documented somewhere.

File: lib/reply.js [hunk 2]
```diff
@@ -126,6 +126,7 @@ Reply.prototype.send = function (payload) {
   }
 
   if (payload instanceof Error || this[kReplyIsError] === true) {
+    this[kReplyIsError] = false
```
Comment (RafaelGSS) [inline lib/reply.js:129]: Why this change? I'm a bit lost since `onErrorHook` and the entire workflow don't manipulate `kReplyIsError`

Comment (sergiz) [reply lib/reply.js:129]: The reason lines in further changes in this file. You already had to turn it to false at two different places. The kReplyIsError is checked only in this section, so it makes sense to me to reset it here instead of chasing it down across the code.

Even

File: lib/contentTypeParser.js (hunk 1 above)
Comment (sergiz) [reply lib/contentTypeParser.js:17]: Good catch.

**Comment by sergiz** [comment]
> 
> 
> I agree with @airhorns, throwing anything but errors it's wrong and should not be supported. Maybe this is a breaking change we should do in v4.
> 

I don't think it's up to fastify to make such judgements. If it was plain wrong then Javascript wouldn't support it in the first place.

**Comment by delvedor** [comment]
@sergiz One of the goals we have as Fastify community is also to educate against bad practices.
If the language does support something it doesn't always mean that is a good thing to do. Especially when we are talking about a 26 years old language :)

File: lib/contentTypeParser.js (hunk 1 above)
Comment (delvedor) [reply lib/contentTypeParser.js]: Can you add a test to cover this file as well?

**Comment by sergiz** [comment]
Any conclusion on this matter? I can also modify the PR to remove this functionality, so only instances of Error would trigger error handling.

**Comment by delvedor** [comment]
@sergiz yes, see https://github.com/fastify/fastify/pull/3200#discussion_r674803949 :)

File: lib/contentTypeParser.js (hunk 1 above)
Comment (sergiz) [reply lib/contentTypeParser.js]: Can you give me an idea how to do that? Isn't this beyond the scope of tests necessary for this PR? I would argue that such tests should then be added for every symbol importation in every file, not just this one.

Comment (delvedor) [reply lib/contentTypeParser.js]: I think it's important to cover this case as well. Especially because there was a typo a nothing failed, which should not happen :D

Comment (sergiz) [reply lib/contentTypeParser.js]: I understand that, but how can you test this? And if you test for this, why not test for all symbols in all files?

Comment (delvedor) [reply lib/contentTypeParser.js]: What I mean is that we should add a test that triggers those errors, so we can validate that the symbol is doing what is expected from it.

Comment (sergiz) [reply lib/contentTypeParser.js]: I added a test that addresses this part.

**Review by delvedor — APPROVED** [review]
LGTM
